doc/man1/qalter.1.in
.\"         OpenPBS (Portable Batch System) v2.3 Software License
.\" 
.\" Copyright (c) 1999-2000 Veridian Information Solutions, Inc.
.\" All rights reserved.
.\" 
.\" ---------------------------------------------------------------------------
.\" For a license to use or redistribute the OpenPBS software under conditions
.\" other than those described below, or to purchase support for this software,
.\" please contact Veridian Systems, PBS Products Department ("Licensor") at:
.\" 
.\"    www.OpenPBS.org  +1 650 967-4675                  sales@OpenPBS.org
.\"                        877 902-4PBS (US toll-free)
.\" ---------------------------------------------------------------------------
.\" 
.\" This license covers use of the OpenPBS v2.3 software (the "Software") at
.\" your site or location, and, for certain users, redistribution of the
.\" Software to other sites and locations.  Use and redistribution of
.\" OpenPBS v2.3 in source and binary forms, with or without modification,
.\" are permitted provided that all of the following conditions are met.
.\" After December 31, 2001, only conditions 3-6 must be met:
.\" 
.\" 1. Commercial and/or non-commercial use of the Software is permitted
.\"    provided a current software registration is on file at www.OpenPBS.org.
.\"    If use of this software contributes to a publication, product, or service
.\"    proper attribution must be given; see www.OpenPBS.org/credit.html
.\" 
.\" 2. Redistribution in any form is only permitted for non-commercial,
.\"    non-profit purposes.  There can be no charge for the Software or any
.\"    software incorporating the Software.  Further, there can be no
.\"    expectation of revenue generated as a consequence of redistributing
.\"    the Software.
.\" 
.\" 3. Any Redistribution of source code must retain the above copyright notice
.\"    and the acknowledgment contained in paragraph 6, this list of conditions
.\"    and the disclaimer contained in paragraph 7.
.\" 
.\" 4. Any Redistribution in binary form must reproduce the above copyright
.\"    notice and the acknowledgment contained in paragraph 6, this list of
.\"    conditions and the disclaimer contained in paragraph 7 in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" 5. Redistributions in any form must be accompanied by information on how to
.\"    obtain complete source code for the OpenPBS software and any
.\"    modifications and/or additions to the OpenPBS software.  The source code
.\"    must either be included in the distribution or be available for no more
.\"    than the cost of distribution plus a nominal fee, and all modifications
.\"    and additions to the Software must be freely redistributable by any party
.\"    (including Licensor) without restriction.
.\" 
.\" 6. All advertising materials mentioning features or use of the Software must
.\"    display the following acknowledgment:
.\" 
.\"     "This product includes software developed by NASA Ames Research Center,
.\"     Lawrence Livermore National Laboratory, and Veridian Information
.\"     Solutions, Inc.
.\"     Visit www.OpenPBS.org for OpenPBS software support,
.\"     products, and information."
.\" 
.\" 7. DISCLAIMER OF WARRANTY
.\" 
.\" THIS SOFTWARE IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND. ANY EXPRESS
.\" OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT
.\" ARE EXPRESSLY DISCLAIMED.
.\" 
.\" IN NO EVENT SHALL VERIDIAN CORPORATION, ITS AFFILIATED COMPANIES, OR THE
.\" U.S. GOVERNMENT OR ANY OF ITS AGENCIES BE LIABLE FOR ANY DIRECT OR INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
.\" OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
.\" EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" 
.\" This license will be governed by the laws of the Commonwealth of Virginia,
.\" without reference to its choice of law rules.
.if \n(Pb .ig Iq
.TH qalter 1B "" Local PBS
.so ../ers/ers.macros
.Iq
.SH NAME
qalter \- alter pbs batch job
.SH SYNOPSIS
qalter
[\-a date_time] [\-A account_string] [\-c interval] [\-e path]
[\-h hold_list] [\-j join] [\-k keep] [\-l resource_list] 
[\-m mail_options] [\-M user_list] [\-N name] [\-o path] [\-p priority]
[\-r c] [\-s c] [\-S path] [\-u user_list] [\-W additional_attributes]
job_identifier...
.SH DESCRIPTION
The
.B qalter
command modifies the attributes of the job or jobs specified
by
.Ar "job_identifier"
on the command line.
Only those attributes listed as options on the command will be modified.  
If any of the specified attributes cannot be modified for a job for any reason, 
none of that job's attributes will be modified.
.LP
The qalter
command accomplishes the modifications by sending a
.I "Modify Job"
batch request to the batch server which owns each job.
.SH OPTIONS
.IP "\-a date_time" 8
Replaces the 
.if !\n(Pb .ig Ig
attribute
.At Execution_Time
.Ig
time at which the job becomes eligible for execution.
The
.Ar date_time
argument syntax is: 
.Ty "[[[[CC]YY]MM]DD]hhmm[.SS]" .
.IP
If the month,
.Ty MM ,
is not specified, it will default to the current month if the specified day
.Ty DD ,
is in the future.  Otherwise, the month will be set to next month.
Likewise, if the day,
.Ty DD ,
is not specified, it will default to today if the time
.Ty hhmm
is in the future.  Otherwise, the day will be set to tomorrow.
.if !\n(Pb .ig Ig
This 
.Ar date_time
will be converted to the integer number of seconds since Epoch that is
equivalent to the local time on the system where the command is being executed.
.Ig
.IP
This attribute can be altered once the job has begun execution, but it will
not take affect until the job is rerun.
.IP "\-A account_string " 8
Replaces the
.if !\n(Pb .ig Ig
the
.At Account_Name
attribute,
.Ig
the account string associated with the job.
.if !\n(Pb .ig Ig
The syntax of the
.Ar account_string
is defined in section 2.7.1. It is interpreted by the server which executes 
the job.
.Ig
.IP
This attribute cannot be altered once the job has begun execution.
.IP "\-c interval" 8
Replaces the
.if !\n(Pb .ig Ig
.At Checkpoint
attribute,
.Ig
the interval at which the job will be checkpointed.
If the job executes upon a host which does not support checkpoint,
this option will be ignored.
.IP
The
.Ar interval
argument is specified as:
.RS
.IP n 4
No checkpointing is to be performed.
.if !\n(Pb .ig Ig
The job's 
.At Checkpoint
attribute is set to the string
.Ty """n""" .
.Ig
.IP s 4
Checkpointing is to be performed only when the server executing the job is
shutdown.
.if !\n(Pb .ig Ig
The job's
.At Checkpoint
attribute is set to the string
.Ty """s""" .
.Ig
.IP c 4
Checkpointing is to be performed at the default minimum cpu time for the
queue from which the job is executing.
.if !\n(Pb .ig Ig
The job's
.At Checkpoint
attribute is set to the string
.Ty """c""" .
.Ig
.IP c=minutes 3
Checkpointing is to be performed at an interval of
.Ar minutes ,
which is the integer number of minutes of CPU time used by the job.
This value must be greater than zero.
If the number is less than the default checkpoint time, the default time
will be used.
.if !\n(Pb .ig Ig
The
.At Checkpoint
attributes is set to the string specified by
.Ar """c=minutes""" .
.Ig
.RE
.IP
This attribute can be altered once the job has begun execution, but the
new value does not take affect until the job is rerun.
.IP "\-e path" 8
Replaces the 
.if !\n(Pb .ig Ig
.At Error_Path
attribute,
.Ig
the path to be used for the standard error stream of the batch job.
The
.Ar path
argument is of the form:
.br
.Ty "\ \ \ \ [hostname:]path_name"
.br
where 
.Ty hostname
is the name of a host to which the file will be returned and
.Ty path_name
is the path
name on that host in the syntax recognized by POSIX 1003.1.
The argument will be interpreted as follows:
.RS
.IP \fBpath_name\fP
Where path_name is not an absolute path name, then the qalter command will
expand the path name relative to the current working directory of the command.
The command will supply the name of the host upon which it is executing for the
.Ar hostname
component.
.IP \fBhostname:path_name\fP
Where path_name is not an absolute path name, then the qalter command will
not expand the path name.  The execution server will expand it relative to
the home directory of the user on the system specified by hostname.
.IP \fBpath_name\fP
Where path_name specifies an absolute path name, then qalter will supply
the name of the host on which it is executing for the
.Ar hostname .
.IP \fBhostname:path_name\fP
Where path_name specifies an absolute path name, the path will be used as
specified.
.RE
.IP
This attribute can be altered once the job has begun execution, but it will
not take affect until the job is rerun.
.IP "\-h hold_list" 8
Updates the 
.if !\n(Pb .ig Ig
.At Hold_Types
attribute,
.Ig
the types of holds on the job.  The
.Ar hold_list
argument is a string of one or more of the following characters:
.RS
.IP u 5
Add the USER type hold.
.IP s 5
Add the SYSTEM type hold if the user has the appropriate level of privilege.
[Typically reserved to the batch administrator.]
.IP o 5
Add the OTHER (or OPERATOR )
type hold if the user has the appropriate level of privilege.
[Typically reserved to the batch administrator and batch operator.]
.IP n 5
Set to none; that is clear the hold types which could be applied with
the users level of privilege.
.RE
.IP
Repetition of characters is permitted, but "n" may not appear in the same
option argument with the other three characters.
This attribute can be altered once the job has begun execution, but the
hold will not take affect until the job is rerun.
.IP "\-j join" 8
Declares which standard streams of the job will be merged together.  The
.Ar join
argument value may be the characters "oe" and "eo", or the
single character "n".
.IP
A argument value of
.Ty oe
directs that the standard output and standard error streams of the job will
be merged, intermixed, and returned as the standard output.
A argument value of
.Ty eo
directs that the standard output and standard error streams of the job will
be merged, intermixed, and returned as the standard error.
.if !\n(Pb .ig Ig
The
.At Join_Path
job attribute is set to the value.
.Ig
.IP
A value of
.Ty n
directs that the two streams will be two separate files.
.if !\n(Pb .ig Ig
The
.At Join_Path
attribute is set to "n".
.Ig
This attribute can be altered once the job has begun execution, but it will
not take affect until the job is rerun.
.IP "\-k keep" 8
Defines which if either of standard output or standard error of the job
will be retained on the execution host.
If set for a stream, this option overrides the path name for that stream.
.IP
The argument is either the single letter "\fBe\fP", "\fBo\fP", or "\fBn\fP",
or one or more of the letters "\fBe\fP" and "\fBo\fP"
combined in either order.
.RS
.IP n 5
No streams are to be retained.
.if !\n(Pb .ig Ig
The 
.At Keep_Files
attribute is set to KEEP_NONE, "n".
.Ig
.IP e 5
The standard error stream is to retained on the execution host.
The  stream will be placed in the home directory of the user under whose
user id the job executed.  The file name will be the default file name
given by:
job_name.\fBe\fPsequence
where \fBjob_name\fP is the name specified for the job, and \fBsequence\fP is
the sequence number component of the job identifier.
.if !\n(Pb .ig Ig
The attribute is set to include KEEP_ERROR, "e".
.Ig
.IP o 5
The standard output stream is to be retained on the execution host.
The  stream will be placed in the home directory of the user under whose
user id the job executed.  The file name will be the default file name
given by: job_name.\fBo\fPsequence
where \fBjob_name\fP is the name specified for the job, and \fBsequence\fP is
the sequence number component of the job identifier.
.if !\n(Pb .ig Ig
The 
.At Output_Path
attribute is set to include KEEP_OUTPUT, "o".
.Ig
.IP eo 5 
Both the standard output and standard error streams will be retained.
.if !\n(Pb .ig Ig
The attribute is set to KEEP_OUTPUT\ |\ KEEP_ERROR.
.Ig
.IP oe 5 
Both the standard output and standard error streams will be retained.
.if !\n(Pb .ig Ig
The attribute is set to KEEP_OUTPUT\ |\ KEEP_ERROR.
.Ig
.RE
.IP
.if !\n(Pb .ig Ig
Repetition of characters is permitted, but "n" may not appear in the same
option argument with the other two characters.
.Ig
This attribute cannot be altered once the job has begun execution.
.IP "\-l resource_list " 8
Modifies the 
.if !\n(Pb .ig Ig
.At Resource_List
attribute, the
.Ig
list of resources that are required by the job.
The 
.Ar Resource_List
argument is in the following syntax:
.br
.Ty resource_name[=[value]][,resource_name[=[value]],...]
.if !\n(Pb .ig Ig
.IP
For each resource listed,
if a resource with the specified name already exist in the jobs resource
attribute, the value for that resource will be updated.
If the named resource does not exist in the job resource attribute,
the resource name and value will be added.
No white space is allowed in the value.
.IP
Because the list of supported resources vary from host to host, the command
will perform no validation of the name or value.  
.Ig
.IP
If a requested modification to a resource would exceed the resource limits
for jobs in the current queue, the server will reject the request.
.IP

CHANGELOG
c - crash     b - bug fix    e - enhancement    f - new feature

2.3.7 Transition
  f - Added new job attribute ATTR_service. Allows job to run as a pseudo
      service (job script accepts start, stop and status commands)
  f - added SIG_SUSPEND/SIG_RESUME capability for pseudo services
      
2.3.7
  e - pbs_mom sisters can now tolerate an explicit group ID instead of only a 
      valid group name. This helps TORQUE be more robust to group lookup 
      failures.
  b - fixed a bug where UNIX domain socket communication was failing when 
      "--disable-privports" was used.
  e - add job exit status as 10th argument to the epilogue script
  c - check filename for NULL to prevent crash
  e - Merged in more logging and NOSIGCHLDMOM capability from Yahoo branch
  e - merged in new log_ext() function to allow more fine grained syslog events, 
      you can now specify severity level. Also added more logging statements
  e - added code to allow compilers to override CLONE_BATCH_SIZE at configure
      time (allows for finer grained control on how arrays are created) (ported
      from Yahoo R2461)
  e - added code which prefixes the severity tag on all log_ext() and log_err()
      messages (ported from Yahoo R2358)
  e - added qmgr option accounting_keep_days, specifies how long to keep
      accounting files.
  e - changed mom config varattr so invoked script returns the varattr name
      and value(s)
  e - improved the performance of pbs_server when submitting large numbers of
      jobs with dependencies defined
  e - added new parameter "log_keep_days" to both pbs_server and pbs_mom.
      Specifies how long to keep log files before they are automatically removed
  e - added qmgr server attribute lock_file, specifies where server lock file
      is located
  e - modified to allow retention of completed jobs across server shutdown
  e - added job_must_report qmgr configuration which says the job must be reported
      to scheduler. Added job attribute "reported". Added PURGECOMP functionality
      which allows scheduler to confirm jobs are reported. Also added -c option
      to qdel. Used to clean up unreported jobs.
  b - Fix so interactive jobs run when using $job_output_file_umask userdefault
  b - Changes to improve the qstat -x XML output and documentation
  b - fix truncated output in qmgr (peter h IPSec+jan n NANCO)
  b - fix so find_resc_entry still works after setting server extra_resc
  b - change so set_jobexid() gets called if JOB_ATR_egroup is not set

2.3.6
  e - in Linux, a pbs_mom will now "kill" a job's task, even if that task can no
      longer be found in the OS processor table. This prevents jobs from getting
      "stuck" when the PID vanishes in some rare cases.
  e - forward-ported change from 2.1-fixes (r2581) (b - reissue job obit even if
      no processes are found)
  b - change back to not sending status updates until we get cluster addr message
      from server, also only try to send hello when the server stream is down.
  b - change pbs_server so log_file_max_size of zero behavior matches 
      documentation
  e - added periodic logging of version and loglevel to help in support
  e - added pbs_mom config option ignvmem to ignore vmem/pvmem limit enforcement
  b - change to correct strtoks that accidentally got changed in astyle 
      formatting

2.3.5
  e - added new init.d scripts for Debian/Ubuntu systems
  b - fixed regression in 2.3.4 release which incorrectly changed soname for 
      libtorque
  b - fixed a bug where TORQUE's exponential backoff for sending messages to the
      MOM could overflow

2.3.4
  b - fixed a bug with RPM spec files due to new pbs_track executable
  b - fixed a bug with "max_report" where jobs not in the Q state were not 
      always being reported to scheduler
  b - fixed bug with new UNIX socket communication when more than one TORQUE 
      instance is running on the same host
  c - fixed a few memory errors due to a spurious comma and some uninitialized 
      memory being allocated
  b - fixed a bug preventing multiple TORQUE servers and TORQUE MOMs from 
      operating properly all from the same host
  f - enabled 'qsub -T' to specify "job type." Currently this will allow a per 
      job prolog/epilog
  f - added a new '-E' option to qstat which allows command-line users to pass 
      "extend" strings via the API
  f - added new max_report queue attribute which will limit the number of Idle 
      jobs, per queue, that TORQUE reports to the scheduler
  e - enhanced logging when a hostname cannot be looked up in DNS
  e - PBS_NET_MAX_CONNECTIONS can now be defined at compile time (via CFLAGS)
  e - modified source code so that all .c and .h files now conform more closely
      to the new CRI format style
  c - fixed segfault when loading job files of an older/incompatible version
  b - fixed a bug where if attempt to send job to a pbs_mom failed due to 
      timeout, the job would indefinitely remain the in 'R' state
  b - fixed a bug where CPU time was not being added up properly in all cases 
      (fix for Linux only) 
  e - pbs_track now allows passing of - and -- options to the a.out argument
  b - qsub now properly interprets -W umask=0XXX as octal umask
  e - allow $HOME to be specified for path
  e - added --disable-qsub-keep-override to allow the qsub -k flag to not 
      override -o -e.
  e - updated with security patches for setuid, setgid, setgroups
  b - fixed correct_ct() in svr_jobfunc.c so we don't crash if we hit COMPLETED
      job
  b - fixed problem where momctl -d 0 showed ConfigVersion twice
  e - if a .JB file gets upgraded pbs_server will back up the original
  b - removed qhold / qrls -h n option since there is no code to support it
  b - set job state and substate correctly when job has a hold attribute and
      is being rerun
  e - fixed several compiler error and warnings for AIX 5.2 systems

2.3.3
  b - fixed bug where pbs_mom would sometimes not connect properly with 

src/cmds/qalter.c
/*
*         OpenPBS (Portable Batch System) v2.3 Software License
*
* Copyright (c) 1999-2000 Veridian Information Solutions, Inc.
* All rights reserved.
*
* ---------------------------------------------------------------------------
* For a license to use or redistribute the OpenPBS software under conditions
* other than those described below, or to purchase support for this software,
* please contact Veridian Systems, PBS Products Department ("Licensor") at:
*
*    www.OpenPBS.org  +1 650 967-4675                  sales@OpenPBS.org
*                        877 902-4PBS (US toll-free)
* ---------------------------------------------------------------------------
*
* This license covers use of the OpenPBS v2.3 software (the "Software") at
* your site or location, and, for certain users, redistribution of the
* Software to other sites and locations.  Use and redistribution of
* OpenPBS v2.3 in source and binary forms, with or without modification,
* are permitted provided that all of the following conditions are met.
* After December 31, 2001, only conditions 3-6 must be met:
*
* 1. Commercial and/or non-commercial use of the Software is permitted
*    provided a current software registration is on file at www.OpenPBS.org.
*    If use of this software contributes to a publication, product, or
*    service, proper attribution must be given; see www.OpenPBS.org/credit.html
*
* 2. Redistribution in any form is only permitted for non-commercial,
*    non-profit purposes.  There can be no charge for the Software or any
*    software incorporating the Software.  Further, there can be no
*    expectation of revenue generated as a consequence of redistributing
*    the Software.
*
* 3. Any Redistribution of source code must retain the above copyright notice
*    and the acknowledgment contained in paragraph 6, this list of conditions
*    and the disclaimer contained in paragraph 7.
*
* 4. Any Redistribution in binary form must reproduce the above copyright
*    notice and the acknowledgment contained in paragraph 6, this list of
*    conditions and the disclaimer contained in paragraph 7 in the
*    documentation and/or other materials provided with the distribution.
*
* 5. Redistributions in any form must be accompanied by information on how to
*    obtain complete source code for the OpenPBS software and any
*    modifications and/or additions to the OpenPBS software.  The source code
*    must either be included in the distribution or be available for no more
*    than the cost of distribution plus a nominal fee, and all modifications
*    and additions to the Software must be freely redistributable by any party
*    (including Licensor) without restriction.
*
* 6. All advertising materials mentioning features or use of the Software must
*    display the following acknowledgment:
*
*     "This product includes software developed by NASA Ames Research Center,
*     Lawrence Livermore National Laboratory, and Veridian Information
*     Solutions, Inc.
*     Visit www.OpenPBS.org for OpenPBS software support,
*     products, and information."
*
* 7. DISCLAIMER OF WARRANTY
*
* THIS SOFTWARE IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND. ANY EXPRESS
* OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
* OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT
* ARE EXPRESSLY DISCLAIMED.
*
* IN NO EVENT SHALL VERIDIAN CORPORATION, ITS AFFILIATED COMPANIES, OR THE
* U.S. GOVERNMENT OR ANY OF ITS AGENCIES BE LIABLE FOR ANY DIRECT OR INDIRECT,
* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
* OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
* EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*
* This license will be governed by the laws of the Commonwealth of Virginia,
* without reference to its choice of law rules.
*/
/*
 *
 * qalter - (PBS) alter batch job
 *
 * Authors:
 *      Terry Heidelberg
 *      Livermore Computing
 *
 *      Bruce Kelly
 *      National Energy Research Supercomputer Center
 *
 *      Lawrence Livermore National Laboratory
 *      University of California
 */

#include "cmds.h"
#include <pbs_config.h>   /* the master config generated by configure */

int main(

  int    argc,  /* I */
  char **argv)  /* I */

  {
  int c;
  int errflg = 0;
  int any_failed = 0;
  char *pc;
  int i;
  int u_cnt, o_cnt, s_cnt, n_cnt;

  struct attrl *attrib = NULL;
  char *keyword;
  char *pdepend;
  char *valuewd;
  time_t after;
  char a_value[80];

  char job_id[PBS_MAXCLTJOBID];

  char job_id_out[PBS_MAXCLTJOBID];
  char server_out[MAXSERVERNAME];
  char rmt_server[MAXSERVERNAME];
  char path_out[MAXPATHLEN + 1];

#define GETOPT_ARGS "a:A:c:e:h:j:k:l:m:M:N:o:p:r:s:S:u:v:W:x:"

  while ((c = getopt(argc, argv, GETOPT_ARGS)) != EOF)
    {
    switch (c)
      {

      case 'a':

        if ((after = cvtdate(optarg)) < 0)
          {
          fprintf(stderr, "qalter: illegal -a value\n");
          errflg++;

          break;
          }

        sprintf(a_value, "%ld",

                (long)after);

        set_attr(&attrib, ATTR_a, a_value);

        break;

      case 'A':

        set_attr(&attrib, ATTR_A, optarg);

        break;

      case 'c':

        while (isspace((int)*optarg))
          optarg++;

        pc = optarg;

        if (strlen(pc) == 1)
          {
          if ((*pc != 'n') && (*pc != 's') && (*pc != 'c'))
            {
            fprintf(stderr, "qalter: illegal -c value\n");

            errflg++;

            break;
            }
          }
        else
          {
          if (strncmp(pc, "c=", 2) != 0)
            {
            fprintf(stderr, "qalter: illegal -c value\n");

            errflg++;

            break;
            }

          pc += 2;

          if (*pc == '\0')
            {
            fprintf(stderr, "qalter: illegal -c value\n");

            errflg++;

            break;
            }

          while (isdigit(*pc))
            pc++;

          if (*pc != '\0')
            {
            fprintf(stderr, "qalter: illegal -c value\n");

            errflg++;

            break;
            }
          }

        set_attr(&attrib, ATTR_c, optarg);

        break;

      case 'e':

        if (prepare_path(optarg, path_out) == 0)
          {
          set_attr(&attrib, ATTR_e, path_out);
          }
        else
          {
          fprintf(stderr, "qalter: illegal -e value\n");

          errflg++;
          }

        break;

      case 'h':

        /* hold */

        while (isspace((int)*optarg))
          optarg++;

        if (strlen(optarg) == 0)
          {
          fprintf(stderr, "qalter: illegal -h value\n");
          errflg++;

          break;
          }

        pc = optarg;

        u_cnt = o_cnt = s_cnt = n_cnt = 0;

        while (*pc)
          {
          if (*pc == 'u')
            u_cnt++;
          else if (*pc == 'o')
            o_cnt++;
          else if (*pc == 's')
            s_cnt++;
          else if (*pc == 'n')
            n_cnt++;
          else
            {
            fprintf(stderr, "qalter: illegal -h value\n");

            errflg++;

            break;
            }

          pc++;
          }

        if (n_cnt && (u_cnt + o_cnt + s_cnt))
          {
          fprintf(stderr, "qalter: illegal -h value\n");

          errflg++;

          break;
          }

        set_attr(&attrib, ATTR_h, optarg);

        break;

      case 'j':

        if ((strcmp(optarg, "oe") != 0) &&
            (strcmp(optarg, "eo") != 0) &&
            (strcmp(optarg, "n")  != 0))
          {
          fprintf(stderr, "qalter: illegal -j value\n");

          errflg++;

          break;
          }

        set_attr(&attrib, ATTR_j, optarg);

        break;

      case 'k':

        /* keep */

        if ((strcmp(optarg, "o")  != 0) &&
            (strcmp(optarg, "e")  != 0) &&
            (strcmp(optarg, "oe") != 0) &&
            (strcmp(optarg, "eo") != 0) &&
            (strcmp(optarg, "n")  != 0))
          {
          fprintf(stderr, "qalter: illegal -k value\n");

          errflg++;

          break;
          }

        set_attr(&attrib, ATTR_k, optarg);

        break;

      case 'l':

        /* resource requirements */

        if (set_resources(&attrib, optarg, TRUE))
          {
          fprintf(stderr, "qalter: illegal -l value\n");

          errflg++;
          }

        break;

      case 'm':

        /* mail event */

        while (isspace((int)*optarg))
          optarg++;

        if (strlen(optarg) == 0)
          {
          fprintf(stderr, "qalter: illegal -m value\n");
          errflg++;

          break;
          }

        if (strcmp(optarg, "n") != 0)
          {
          pc = optarg;

          while (*pc)
            {
            if ((*pc != 'a') && (*pc != 'b') && (*pc != 'e'))
              {
              fprintf(stderr, "qalter: illegal -m value\n");

              errflg++;

              break;
              }

            pc++;
            }
          }

        set_attr(&attrib, ATTR_m, optarg);

        break;

      case 'M':

        /* mail destination */

        if (parse_at_list(optarg, FALSE, FALSE))
          {
          fprintf(stderr, "qalter: illegal -M value\n");
          errflg++;

          break;
          }

        set_attr(&attrib, ATTR_M, optarg);

        break;

      case 'N':

        /* name */

        if (check_job_name(optarg, 1) == 0)
          {
          set_attr(&attrib, ATTR_N, optarg);
          }
        else
          {
          fprintf(stderr, "qalter: illegal -N value\n");

          errflg++;
          }

        break;

      case 'o':

        /* output */

        if (prepare_path(optarg, path_out) == 0)
          {
          set_attr(&attrib, ATTR_o, path_out);
          }
        else
          {
          fprintf(stderr, "qalter: illegal -o value\n");

          errflg++;
          }

        break;

      case 'p':

        /* priority */

        while (isspace((int)*optarg))
          optarg++;

        pc = optarg;

        if ((*pc == '-') || (*pc == '+'))
          pc++;

        if (strlen(pc) == 0)
          {
          fprintf(stderr, "qalter: illegal -p value\n");

          errflg++;

          break;
          }

        while (*pc != '\0')
          {
          if (!isdigit(*pc))
            {
            fprintf(stderr, "qalter: illegal -p value\n");

            errflg++;

            break;
            }

          pc++;
          }

        i = atoi(optarg);

        if ((i < -1024) || (i > 1023))
          {
          fprintf(stderr, "qalter: illegal -p value\n");

          errflg++;

          break;
          }

        set_attr(&attrib, ATTR_p, optarg);

        break;

      case 'r':

        /* rerun */

        if (strlen(optarg) != 1)
          {
          fprintf(stderr, "qalter: illegal -r value\n");

          errflg++;

          break;
          }

        if ((*optarg != 'y') && (*optarg != 'n'))
          {
          fprintf(stderr, "qalter: illegal -r value\n");

          errflg++;

          break;
          }

        set_attr(&attrib, ATTR_r, optarg);

        break;

      case 's':

        /* service */

        if (strlen(optarg) != 1)
          {
          fprintf(stderr, "qalter: illegal -s value\n");

          errflg++;

          break;
          }

        if ((*optarg != 'y') && (*optarg != 'n'))
          {
          fprintf(stderr, "qalter: illegal -s value\n");

          errflg++;

          break;
          }

        set_attr(&attrib, ATTR_service, optarg);

        break;

      case 'S':

        if (parse_at_list(optarg, TRUE, TRUE))
          {
          fprintf(stderr, "qalter: illegal -S value\n");

          errflg++;

          break;
          }

        set_attr(&attrib, ATTR_S, optarg);

        break;

      case 'u':

        if (parse_at_list(optarg, TRUE, FALSE))
          {
          fprintf(stderr, "qalter: illegal -u value\n");

          errflg++;

          break;
          }

        set_attr(&attrib, ATTR_u, optarg);

        break;

      case 'v':

        set_attr(&attrib, ATTR_v, optarg);

        break;

      case 'W':

        while (isspace((int)*optarg))
          optarg++;

        if (strlen(optarg) == 0)
          {
          fprintf(stderr, "qalter: illegal -W value\n");

          errflg++;

          break;
          }

        i = parse_equal_string(optarg, &keyword, &valuewd);

        while (i == 1)
          {
          if (strcmp(keyword, ATTR_depend) == 0)
            {
            pdepend = malloc(PBS_DEPEND_LEN);

            if ((pdepend == NULL) ||
                 parse_depend_list(valuewd,pdepend,PBS_DEPEND_LEN))
              {
              fprintf(stderr, "qalter: illegal -W value\n");

              errflg++;

              break;
              }

            valuewd = pdepend;
            }
          else if (!strcmp(keyword,ATTR_stagein))
            {
            if (parse_stage_list(valuewd))
              {
              fprintf(stderr, "qalter: illegal -W value\n");

              errflg++;

              break;
              }
            }
          else if (strcmp(keyword,ATTR_stageout) == 0)
            {
            if (parse_stage_list(valuewd))
              {
              fprintf(stderr,"qalter: illegal -W value\n");

              errflg++;

              break;
              }
            }
          else if (strcmp(keyword,ATTR_g) == 0)
            {
            if (parse_at_list(valuewd,TRUE,FALSE))
              {
              fprintf(stderr,"qalter: illegal -W value\n");

              errflg++;

              break;
              }
            }

          set_attr(&attrib, keyword, valuewd);

          i = parse_equal_string(NULL, &keyword, &valuewd);
          }

        if (i == -1)
          {
          fprintf(stderr, "qalter: illegal -W value\n");

          errflg++;
          }

        break;

      case 'x':

        /* exec_hosts */

        set_attr(&attrib, ATTR_exechost, optarg);

        break;

      case '?':

      default:

        errflg++;

        break;
      }  /* END switch(c) */
    }    /* END while(c = getopt()) */

  if (errflg || (optind == argc))
    {
    static char usage[] = "usage: qalter \
        [-a date_time] [-A account_string] [-c interval] [-e path] \n\
        [-h hold_list] [-j y|n] [-k keep] [-l resource_list] [-m mail_options] \n\
        [-M user_list] [-N jobname] [-o path] [-p priority] [-r y|n] [-s y|n] \n\
        [-S path] [-u user_list] [-v variable_list] [-W dependency_list] \n\
        [-x exec_host] job_identifier...\n";

    fprintf(stderr, "%s", usage);

    exit(2);
    }

  for (;optind < argc;optind++)
    {
    int connect;
    int stat = 0;
    int located = FALSE;

    strcpy(job_id, argv[optind]);

    if (get_server(job_id, job_id_out, server_out))
      {
      fprintf(stderr, "qalter: illegally formed job identifier: %s\n",
              job_id);

      any_failed = 1;

      continue;
      }

cnt:

    connect = cnt2server(server_out);

    if (connect <= 0)
      {
      fprintf(stderr, "qalter: cannot connect to server %s (errno=%d) %s\n",
              pbs_server,
              pbs_errno,
              pbs_strerror(pbs_errno));

      any_failed = pbs_errno;

      continue;
      }

    stat = pbs_alterjob(connect, job_id_out, attrib, NULL);

    if ((stat != 0) && (pbs_errno != PBSE_UNKJOBID))
      {
      prt_job_err("qalter", connect, job_id_out);

      any_failed = pbs_errno;
      }
    else if ((stat != 0) && (pbs_errno == PBSE_UNKJOBID) && !located)
      {
      located = TRUE;

      if (locate_job(job_id_out, server_out, rmt_server))
        {
        pbs_disconnect(connect);

        strcpy(server_out, rmt_server);

        goto cnt;
        }

      prt_job_err("qalter", connect, job_id_out);

      any_failed = pbs_errno;
      }  /* END else if (stat && ...) */

    pbs_disconnect(connect);
    }  /* END for (optind) */

  exit(any_failed);
  }  /* END main() */

/* END qalter.c */
